### ISSUE TITLE
Include chi2 of excluded datasets in training reports

### ISSUE THREAD

**Issue by scarlehoff** [NNPDF/nnpdf#1707]
People in the mailing list have manifested their wish to have the chi2 of excluded datasets in the training reports. I think it could be a nice addition (but I would have it as a separate table).

An example of such a table is given by @enocera here https://vp.nnpdf.science/o7dD198iRsSv_UFS6bwPAw==

Note that this is not trivial since it can only be done in some cases (for instance, if the theory choices are different in both reports, such a table cannot be constructed in an unambiguous way).

**Comment by enocera** [comment]
> Note that this is not trivial since it can only be done in some cases (for instance, if the theory choices are different in both report, such a table cannot be constructed in an unambiguous way).

Exactly. So do we really need do do that in the reports given that, as @Zaharid says, it's easy enough to write a vp runcard that does this job for you outside the reports? Life is short...

**Comment by stefanoforte** [comment]
Actually,  I don';t think that an extra table like the one produced by @enocera in this example is what we need. We simply need the standard default plots and numbers that are included in the TR to be present for the union rather the interesection of the datasets whenever comparing two fits that use a different dataset. I'd be happy enough if this option was only available for fits that differ only in dataset and not in other aspects - in which case there is no ambiguity and the FK tables are surely available. Indeed whenever we do fits to different datsets what we want is to have the chi2 histograms and tables and data-theory comparisons exactly with the same layout and format in the two fits that are being compared.

**Comment by enocera** [comment]
@stefanoforte Do you really mean the union of the data sets of the two fits that are compared in a vp report? Suppose that we compare a FT DIS-only fit and a LHC-only fit. The union will be made of these two data sets, which are a subset of the NNPDF4.0 data set. But perhaps in this case you'd like to see the chi2 for data sets that are not included in either fit, e.g.  HERA?

**Comment by stefanoforte** [comment]
@enocera , if I compare something I wish to compare what I am comparing. If I wanted to compare to the global fit that's what I would do. If we start including stuff outside the comparison that it is being made then it truly becomes fully ambiguous. I'd say instead that when making a comparison the only really meaningful choice is whether we want to show plots for the union or the intersection. To me the union sounds much more natural because that's the whole point of the comparison.

**Comment by stefanoforte** [comment]
@enocera indeed that was exactly my point: in the table we have NaNs and in the plots we move all datasets not included to the top with just one of the two histo bars showing. So there's a sense in which we always show the union already, except that we do it incompletely.  I'd say that what we want is to make a full comparison all the time.

**Comment by Zaharid** [comment]
The tables compute the intersection rather than  the union. The nans are an artefact of how the tables get joined. The fundamental problem is that for the intersection we can conceptually "copy" the configuration the fit was ran with, while for the union we would have to "invent" it in the most general case.

The way forward to have this in the report would be to somehow detect the restricted situation where we don't have to invent much. Detecting programmatically that fits only differ in the dataset is doable but tricky: Does the possible difference t0 count? Do we  compare all the cuts in all the dataset? What do we do about different cfactors? An so on.

In that sense, having an external runcard without the desired plots and tables would be both more flexible and easier.

**Comment by stefanoforte** [comment]
Buy why do we have to detect it programatically? Comparing fits that only differ by dataset is what we do half the time. Can't we just have it as a declarative option? The report that we get when comparing fits that only differ by dataset includes the union. IMHO it is a commom enough situation that it deserves a dedicated option. If any of the required pieces of information is missing we fall back on the usual report. The external runcard solution disregards the fact that whenever we run dataset variations we typically do dozens at a time (just look at the wiki pages for IC fits, MHOU fits etc etc). Again, this is a very commoin situation in which we typically do of order 10-20 fits in a row.

**Comment by Zaharid** [comment]
Something actually nice would be to have fit runcards that look like

```
baseline: <existing fit>

modified_settings:
    ...
```

Which would then copy all the settings from the baseline and afford extra functionality like these plots. But would be the deluxe option in terms of adding stuff to the code, in addition to being its own can of worms in terms of random seeds, preprocessing exponents and so on.

A simpler, but more limited option would be to have some kind of annotation in the runcard for comparefits only. One would have to think what the annotation is so as to not make it too clunky.

Yet another option would be to specify the relationship somehow at the time comparefits is executed. But IMO that is clearly in the territory of diminishing returns as as that point on can simply copy the [runcard](https://github.com/NNPDF/nnpdf/tree/master/validphys2/src/validphys/comparefittemplates) and edit by hand as needed.

**Comment by stefanoforte** [comment]
indeed, the deluxe option looks potentially complicated given that we also do (though clearly less frequently) methodology variations. Personally I'd be happy with the simpler limited option.

**Comment by Zaharid** [comment]
I am thinking that unless we want to use this for something else, contrary to what I said previously, it is best to have this as a command line option in vp.

The rough semantics would be that the we have `--dataset-join={intersection, union, left, right}` (or maybe `{intersection, union, current, reference}`), where the default is intersection.

`intesection` would behave similar to now, `current` would add also a comparison for the datasets that are in the current fit and not in the reference reference, `reference` would add to `intersection` a comparison for the datsets that are in `reference` but not `current`, and finally `union` would add both of these. 

Then effectively we have up to three namespacepaces for dataset related actions, depending on the selected option: The one with the intersection (that we have now), the one for the  datasets that are in the baseline but not the reference, and the one in the reference but not the baseline.  For the right and left cases, we would copy all the settings `from_` the fit where the dataset is available and only loop over the pdf between basline and reference to produce the comparison. 

The namespaces would be split in the report so as to decrease the possible confusion (as in principle these are all computing different things).

This is not the most general thing but it would succeed always and is good enough for the case where we only vary the dataset with reasonably clear semantics. 

The problems with the in-runcard key are that people will forget it in the common case and also that the semantics would be cumbersome (e.g. how it behaves when the fit is a "reference" versus "current").

Implementing this (in any version) would be a  fair amount of work.

**Comment by Zaharid** [comment]
A drawback of the last idea would be the potential to output fairly nonsensical things, like for example when comparing NLO and NNLO fits. I guess that's why the options would be off by default and left to the author.

**Comment by Zaharid** [comment]
> What would be the benefit of `current` and `reference` on top of `union`? It seems very limited to me and probably not really worth cluttering the code with.

The benefits would be very small, but on the other hand the cost would be as well. I am thinking one the very niche case where one of the comparisons doesn't work (e.g. fktable not in theory) but one still wants to see the other comparison. The implementation would need to handle these options for the union so it would be a question of whether they should be in the user interface or are too niche.

> 
> > Then effectively we have up to three namespacepaces for **dataset related actions**
> 
> Though I guess in the TR we only want to use this for the chi2 tables and histograms, not elsewhere (or do we?). So we'd have to use two data configurations to pruduce a TR (which is fine of course).

In my mind, the discussion is about the chi2 tables and plots, and the data theory comparisons. The rest would work the same.

**Comment by scarlehoff** [comment]
This has now been fixed by #2387
